ldap2pg/manager.py

```diff
@@ -55,6 +55,44 @@ def query_ldap(self, base, filter, attributes, scope):
                 raise UserError(message)
 
             entries.append(lower_attributes(entry))
+
+        return entries
+
+    def query_ldap(self, base, filter, attributes, joins, scope):
+        entries = self._query_ldap(base, filter, attributes, scope)
+        sub_attrs_cache = dict()
+        for entry in entries:
+            for attr, values in list(entry[1].items()):
+                values_with_attrs = []
+                for value in values:
+                    join = joins.get(attr)
+                    if join is None:
+                        values_with_attrs.append((value, dict()))
+                        continue
+
+                    sub_attrs = sub_attrs_cache.get((attr, value))
+                    if sub_attrs:
+                        values_with_attrs.append((value, dict(sub_attrs)))
+                        continue
+
+                    sub_attrs = {'dn': [value]}
+
+                    if not join['attributes']:
+                        values_with_attrs.append((value, sub_attrs))
+                        sub_attrs_cache[(attr, value)] = sub_attrs
+                        continue
+
+                    join = dict(join, base=value)
+                    try:
+                        join_values = self._query_ldap(**join)
+                        if join_values:
+                            sub_attrs.update(join_values[0][1])
+                            values_with_attrs.append((value, sub_attrs))
+                            sub_attrs_cache[(attr, value)] = sub_attrs
+                    except UserError as e:
+                        logger.warn('Ignoring %s: %s' % (value, e))
```

Comment:
A detail this line should be:

``` python
logger.warning('Ignoring %s: %s', value, e)
```

`logger.warn` is deprecated by `logging`. Formatting should be done by `logging`.

Reply:
Right, sorry.